gobject-base/src/object.rs

```diff
@@ -86,43 +89,93 @@ where
         unimplemented!()
     }
 
-    unsafe fn get_instance(&self) -> *mut InstanceStruct<Self>;
+    unsafe fn get_instance(&self) -> *mut Self::InstanceStructType;
+
 
-    fn get_impl(&self) -> &Self::ImplType {
+    fn get_impl(&self) -> &Self::ImplType{
         unsafe { (*self.get_instance()).get_impl() }
     }
 
-    unsafe fn get_class(&self) -> *const ClassStruct<Self> {
+    unsafe fn get_class(&self) -> *const ClassStruct<Self>{
         (*self.get_instance()).get_class()
     }
 }
 
 #[macro_export]
 macro_rules! object_type_fns(
     () => {
-        unsafe fn get_instance(&self) -> *mut InstanceStruct<Self> {
+        unsafe fn get_instance(&self) -> *mut Self::InstanceStructType {
             self.to_glib_none().0
         }
     }
 );
 
-#[repr(C)]
-pub struct InstanceStruct<T: ObjectType> {
-    pub parent: T::GlibType,
-    pub imp: ptr::NonNull<T::ImplType>,
-    pub panicked: AtomicBool,
-}
 
-impl<T: ObjectType> InstanceStruct<T> {
-    pub fn get_impl(&self) -> &T::ImplType {
-        unsafe { self.imp.as_ref() }
-    }
+pub trait Instance<T: ObjectType>
+{
+    fn parent(&self) -> &T::GlibType;
 
-    pub unsafe fn get_class(&self) -> *const ClassStruct<T> {
-        *(self as *const _ as *const *const ClassStruct<T>)
-    }
+    fn get_impl(&self) -> &T::ImplType;
+    fn get_impl_ptr(&self) -> *mut T::ImplType;
```

Comment:
Why do we need both?

Reply:
There's only one spot where ```get_impl_ptr``` is and that's in the drop call of ```finalize<T: ObjectType>```.  I did not want to expose the NonNull instance itself further.
I would remove this if I knew how to get from a &T to a *mut T.